### PR TITLE
Protocol v5.2 with TLS transport

### PR DESCRIPTION
**Pull request by daniel-lerch**
This pull request implements the current draft for the upcoming protocol v5.2 and comes with several other improvements:
- TLS based transport protocol
- Dependency Injection everywhere
- Send Queue
- Asynchronous message delivery

### CODE REVIEW

File: src/SkynetServer/Network/MessageHandler.cs [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    internal class MessageHandler<T> : PacketHandler<T> where T : ChannelMessage
+    {
+        public sealed override async ValueTask Handle(T packet)
+        {
+            if (!packet.MessageFlags.AreValid(packet.RequiredFlags, packet.AllowedFlags))
+                throw new ProtocolException($"Invalid MessageFlags{packet.MessageFlags} for content packet ID {packet.Id}");
+
+            if (packet.MessageFlags.HasFlag(MessageFlags.Unencrypted)
+                && await Validate(packet).ConfigureAwait(false) != MessageSendStatus.Success)
+                return; // Not all messages can be saved, some return MessageSendError other than Success
```
Comment (Twometer) [inline src/SkynetServer/Network/MessageHandler.cs]:
```suggestion
                return; // Not all messages can be saved, some return MessageSendStatus other than Success
```

File: src/SkynetServer/Network/Packets/P1CDirectChannelCustomization.cs [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+using SkynetServer.Network.Attributes;
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace SkynetServer.Network.Packets
+{
+    [Packet(0x1C, PacketPolicies.Duplex)]
+    [MessageFlags(MessageFlags.Loopback)]
+    internal sealed class P1CDirectChannelCustomization : ChannelMessage
```
Comment (Twometer) [inline src/SkynetServer/Network/Packets/P1CDirectChannelCustomization.cs]: Do we need empty packet classes for packets the server can't handle due to encryption?

File: src/SkynetServer/Network/Packets/P27ProfileImage.cs [hunk 3]
```diff
@@ -1,30 +1,26 @@
...
-using System.Threading.Tasks;
-using VSL;
 
 namespace SkynetServer.Network.Packets
 {
-    [Message(0x27, PacketPolicy.Duplex)]
-    [RequiredFlags(MessageFlags.FileAttached)]
-    [AllowedFlags(MessageFlags.Unencrypted | MessageFlags.FileAttached)]
-    internal sealed class P27ProfileImage : P0BChannelMessage
+    [Packet(0x27, PacketPolicies.Duplex)]
```
Comment (Twometer) [inline src/SkynetServer/Network/Packets/P27ProfileImage.cs]: [OT] I am so far down the review page that syntax highlighting stopped working [/OT]

File: src/SkynetServer/Services/MessageInjectionService.cs [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+using SkynetServer.Model;
+using SkynetServer.Network;
+using SkynetServer.Network.Model;
+using SkynetServer.Network.Packets;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace SkynetServer.Services
```
Comment (Twometer) [inline src/SkynetServer/Services/MessageInjectionService.cs]: [OT] Now, it not only doesn't syntax-highlight, but also no longer loads diffs [/OT]

File: src/SkynetServer/Network/Packets/P1CDirectChannelCustomization.cs (hunk 2 above)
Comment (daniel-lerch) [reply src/SkynetServer/Network/Packets/P1CDirectChannelCustomization.cs]: Yes, we need these empty classes in order to enforce `MessageFlags`.

File: src/SkynetServer/Network/Handlers/P04DeleteAccountHandler.cs [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+                .ToArrayAsync().ConfigureAwait(false);
+            Session[] sessions = await Database.Sessions.AsQueryable()
+                .Where(s => s.AccountId == Client.AccountId)
+                .ToArrayAsync().ConfigureAwait(false);
+            Database.MailConfirmations.RemoveRange(mailConfirmations);
+            Database.Sessions.RemoveRange(sessions);
+            await Database.SaveChangesAsync().ConfigureAwait(false);
+
+            // Kick all sessions
+            // TODO: Evaluate better options than waiting here which opens the door for dead locks
```
[the remark this answers is not in the document]
Comment (daniel-lerch) [reply src/SkynetServer/Network/Handlers/P04DeleteAccountHandler.cs]: It's very unlikely that one user deletes its account on two devices simultaneously. Therefore, I will fix this problem in another pull request.

**Comment by daniel-lerch** [comment]
Azure Pipelines builds fail since the 20200316.1 VM update for an unknown reason. In Visual Studio on Windows with MariaDB all tests pass.

**Pull request merged**